### PR TITLE
swarm: enable p2p/discovery and disable dynamic dialling

### PR DESCRIPTION
**Pull request by nonsense**
This PR is re-enabling p2p discovery, but disabling the dynamic dialling functionality - the reason why we didn't want to have the p2p discovery in the first place.

The reason for this is - the `hive` protocol and the `Kademlia` table implementation on Swarm side does not update the underlay addresses of peers. More specifically `Kademlia.Register` does not update the underlay addresses ever, but `Kademlia.On` and `Kademlia.Off` do. This means that sometimes we are left with outdated destinations for nodes in our state store.

The `p2p` package handles that with:

```
    err := t.dial(srv, t.dest)
    if err != nil {
        log.Trace("Dial error", "task", t, "err", err)
        // Try resolving the ID of static nodes if dialing failed.
        if _, ok := err.(*dialError); ok && t.flags&staticDialedConn != 0 {
            if t.resolve(srv) {
                t.dial(srv, t.dest)
            }
        }
    }
```

Namely, if `dialling` fails, it tries to `resolve` an up-to-date destination for the node - something that is only done when `discovery` is enabled.

Bottom line: `hive` protocol relies heavily on the discovery functionality in the `p2p` package, so for now it is not wise to disable discovery on `p2p`.

Addresses: https://github.com/ethersphere/go-ethereum/issues/1268

-----

I also considered solving this directly on Swarm side, but the `hive` protocol implementation is not straight-forward and it is not clear how to determine what an up-to-date / accurate node destination is. What happens is that when a new node starts up and loads an out-dated node destination, it might receive a `hive` message with a different destination for a given node, and it can't determine which one is correct, without trying them both. We don't have functionality to do that, and implementing this seems like a major change in the `hive` protocol.

### CODE REVIEW

File: swarm/network/kademlia.go [hunk 1]
```diff
@@ -149,10 +150,22 @@ func (k *Kademlia) Register(peers ...*BzzAddr) error {
...
+				log.Trace("registering new peer", "addr", p)
 				// insert new offline peer into conns
 				return newEntry(p)
 			}
-			// found among known peers, do nothing
+
+			e := v.(*entry)
+
+			// if underlay address is different, still add
+			if bytes.Compare(e.BzzAddr.UAddr, p.UAddr) != 0 {
```
Comment (janos) [inline swarm/network/kademlia.go]: Lint complains about this:

```
swarm/network/kademlia.go:161:7:warning: should use !bytes.Equal(e.BzzAddr.UAddr, p.UAddr) instead (S1004) (gosimple)
```

**Review by janos — APPROVED** [review]
With lint warning fix, LGTM.

**Comment by zelig** [comment]
@nonsense I would simply implement a timestamp on the record. 
If newer than the underlay you have for peer, you switch it.

File: swarm/api/inspector.go [hunk 2]
```diff
@@ -38,6 +39,14 @@ func (inspector *Inspector) Hive() string {
 	return inspector.hive.String()
 }
 
+func (inspector *Inspector) ListKnown() []string {
```
Comment (zelig) [inline swarm/api/inspector.go:42]: comment exported func?

File: swarm/network/kademlia.go [hunk 3]
```diff
@@ -149,10 +150,22 @@ func (k *Kademlia) Register(peers ...*BzzAddr) error {
...
+				log.Trace("registering new peer", "addr", p)
 				// insert new offline peer into conns
 				return newEntry(p)
 			}
-			// found among known peers, do nothing
+
+			e := v.(*entry)
+
+			// if underlay address is different, still add
+			if !bytes.Equal(e.BzzAddr.UAddr, p.UAddr) {
```
[the remark this answers is not in the document]
Comment (nonsense) [reply swarm/network/kademlia.go:161]: how exactly would you implement a timestamp in a distributed environment? we can't trust time coming from other nodes as far as I know.

**Comment by nonsense** [comment]
> @nonsense I would simply implement a timestamp on the record.
> If newer than the underlay you have for peer, you switch it.

could you elaborate on that? nodes don't have access to a global clock, we have no way of trusting a timestamp coming from peers... we might as well just randomly select one of the destinations.

**Comment by nonsense** [comment]
@zelig I saw your comments too late, as this already got two approvals upstream. Let's discuss further and decide how to improve. I am not sure how the timestamp you propose can be a solution to this, maybe I am missing something.
